src/tacacs_plus_fmt_plug.c

```diff
@@ -0,0 +1,337 @@
+/*
+ * Format for cracking TACACS+ hashes.
+ *
+ * https://insinuator.net/2015/06/tacacs-module-for-loki/
+ *
+ * This software is Copyright (c) 2015, Daniel Mende <dmende [at] ernw.de> and
+ * Copyright (c) 2017, Dhiru Kholia <dhiru [at] openwall.com>, and it is hereby
+ * released to the general public under the following terms:
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *
+ *  * Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ *
+ *  * Redistributions in binary form must reproduce the above
+ *    copyright notice, this list of conditions and the following disclaimer
+ *    in the documentation and/or other materials provided with the
+ *    distribution.
+ *
+ *  * Neither the name of the  nor the names of its
+ *    contributors may be used to endorse or promote products derived from
+ *    this software without specific prior written permission.
+ *
+ *    THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ *    "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ *    LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ *    A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ *    OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ *    SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ *    LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ *    DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ *    THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ *    (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ *    OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#if FMT_EXTERNS_H
+extern struct fmt_main fmt_tacacsplus;
+#elif FMT_REGISTERS_H
+john_register_one(&fmt_tacacsplus);
+#else
+
+#include <string.h>
+#ifdef _OPENMP
+#include <omp.h>
+#ifndef OMP_SCALE
+#define OMP_SCALE               2048
+#endif
+#endif
+
+#include "formats.h"
+#include "misc.h"
+#include "common.h"
+#include "params.h"
+#include "options.h"
+#include "md5.h"
+#include "johnswap.h"
+#include "memdbg.h"
+
+#define FORMAT_LABEL            "tacacs-plus"
+#define FORMAT_NAME             "TACACS+"
+#define FORMAT_TAG              "$tacacs-plus$"
+#define TAG_LENGTH              (sizeof(FORMAT_TAG) - 1)
```

Comment:
We make a lot of use of this `sizeof(string constant)` pattern already, so it's certainly OK for this PR/commit, but IIRC it was until some point a GNU C extension and is probably part of some newer C standard now. We might want to figure that out and target a consistent minimum C standard. I am not sure if C99 already permitted this or not yet. (I refrained from using this construct in my code so far, so didn't bother checking.) @magnumripper?

Reply:
I wasn't aware of this but never heard of any problem with it. Not sure how certain one can be that a strlen would be optimized away. A hard-coded figure is error-prone for this thing (almost every new format starts out as a copy of some similar format nowadays).